cumulus/parachains/runtimes/bridge-hubs/test-utils/src/test_cases/helpers.rs

```diff
@@ -389,6 +494,112 @@ fn execute_and_verify_calls<Runtime: frame_system::Config>(
 	}
 }
 
+/// Trait for ensuring XCM message delivery and retrieving messages for a given location
+///
+/// Used to abstract over different message delivery mechanisms like HRMP channels
+/// and message queues.
+pub trait EnsureDeliveryAndMessage {
+	/// Sets up any required message delivery infrastructure for the given location.
+	fn ensure_delivery_for(location: &Location) -> Result<(), XcmError>;
+
+	/// Retrieves any XCM messages ready to be delivered to the given location.
+	fn get_xcm_to_deliver_for(location: &Location) -> Option<VersionedXcm<()>>;
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(8)]
+impl EnsureDeliveryAndMessage for Tuple {
+	fn ensure_delivery_for(location: &Location) -> Result<(), XcmError> {
+		for_tuples!( #(
+			if let Err(e) = Tuple::ensure_delivery_for(location) {
+				return Err(e)
+			}
+		)* );
+		Ok(())
+	}
+
+	fn get_xcm_to_deliver_for(location: &Location) -> Option<VersionedXcm<()>> {
+		for_tuples!( #(
+			if let Some(xcm) = Tuple::get_xcm_to_deliver_for(location) {
+				return Some(xcm);
+			}
+		)* );
+		None
+	}
+}
+
+/// An implementation of `EnsureDeliveryAndMessage` for `Here` location.
+/// - reads XCM from the `pallet-message-queue`
+pub struct ToMessageQueueDelivery<Runtime>(PhantomData<Runtime>);
+impl<Runtime: pallet_message_queue::Config> EnsureDeliveryAndMessage
+	for ToMessageQueueDelivery<Runtime>
+where
+	pallet_message_queue::MessageOriginOf<Runtime>: for<'a> TryFrom<&'a Location>,
+{
+	fn ensure_delivery_for(_location: &Location) -> Result<(), XcmError> {
+		Ok(())
+	}
+
+	fn get_xcm_to_deliver_for(location: &Location) -> Option<VersionedXcm<()>> {
+		if !matches!(location.unpack(), (0, [])) {
+			return None;
+		}
+
+		let Ok(origin): Result<pallet_message_queue::MessageOriginOf<Runtime>, _> =
+			location.try_into()
+		else {
+			return None;
+		};
+		// read page index from state
+		// TODO: FAIL-CI - (Serban) if 0 is ok, we remove the line bellow
+		// let page_index = pallet_message_queue::BookStateFor::<Runtime>::get(origin).begin;
+		let page_index = 0;
+		// get page
+		let Some(page) = pallet_message_queue::Pages::<Runtime>::get(origin, page_index) else {
+			return None;
+		};
+		// find/peek first unprocessed message
+		// TODO: FAIL-CI - (Serban) - how?
+		page.peek_first()
+			.map(|msg| VersionedXcm::<()>::decode(&mut &msg[..]).expect("valid XCM"))
```

Comment:
Yes, this is the right approach. `page.peek_first()` will return the first unprocessed message on the page. But this will work only if we enqueue just 1 message. If we enqueue more, `page.peek_first()` will always return the first unprocessed message and we'll never reach the following ones.